### PR TITLE
Feature #3523:  Light source on magic projectiles

### PR DESCRIPTION
**Pull request by Aussiemon**
[Feature #3523](https://bugs.openmw.org/issues/3523)

Attached light to magic bolt projectiles.

[Thread discussion of this pull request](https://forum.openmw.org/viewtopic.php?f=6&t=3972)

### CODE REVIEW

File: apps/openmw/mwworld/projectilemanager.cpp [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 #include <iomanip>
 
+#include <osg/Light>
+#include <osg/LightSource>
```
Comment (scrawl) [inline apps/openmw/mwworld/projectilemanager.cpp]: Correct include is components/sceneutil/lightmanager.hpp

File: apps/openmw/mwworld/projectilemanager.cpp [hunk 2]
```diff
@@ -136,7 +138,7 @@ namespace MWWorld
     };
 
 
-    void ProjectileManager::createModel(State &state, const std::string &model, const osg::Vec3f& pos, const osg::Quat& orient, bool rotate, std::string texture)
+    void ProjectileManager::createModel(State &state, const std::string &model, const osg::Vec3f& pos, const osg::Quat& orient, bool rotate, bool isMagic, std::string texture)
```
Comment (scrawl) [inline apps/openmw/mwworld/projectilemanager.cpp]: isMagic should be renamed to createLight to reflect the variable's real purpose.

**Comment by Aussiemon** [comment]
Retrieval of colors from the relevant magic effects is complete. Testing on several spells shows that the blended diffusion values match the projectile colors in vanilla Morrowind.

**Comment by scrawl** [comment]
> Retrieval of colors from the relevant magic effects is complete. Testing on several spells shows that the blended diffusion values match the projectile colors in vanilla Morrowind.

That sounds very.. scientific ;)

**Comment by MiroslavR** [comment]
@scrawl, can we agree not to use C-style casts anymore?

**Comment by scrawl** [comment]
Oh, I didn't notice the casts.

A ```static_cast<float>``` would be preferred over ```(float)```, but these are a very minor issue that I'm not going to bother bringing up in review.

Regarding this one:
<pre>
int numberOfEffects = ((MagicBoltState&)state).mEffects.mList.size();
</pre>
Is a problem, the function will crash if for some reason createLight is set to true for a non-magic projectile.

@Aussiemon: could you make a new PR updating the code to not use C-style casts, and remove the MagicBoltState cast? It's probably easiest to just compute the light color before hand and then pass it to createModel.

**Comment by MiroslavR** [comment]
> A static_cast<float> would be preferred over (float), but these are a very minor issue that I'm not going to bother bringing up in review.

I was primarily referring to non-trivial types. The ability to fall back to reinterpret_cast sounds like an accident waiting to happen. C++ casts express their intent much more clearly.

**Comment by scrawl** [comment]
Yes, that is my stance. For non-trivial types, we should always use C++ casts.

In this case, the cast shouldn't be there in the first place though.

**Comment by MiroslavR** [comment]
> In this case, the cast shouldn't be there in the first place though.

Agreed.

**Comment by Aussiemon** [comment]
Thanks for the review of my work. I'll definitely take this into account in the future. 

I've created a [pull request](https://github.com/OpenMW/openmw/pull/1151) with the recommended changes to casts. If there's a better way to make these fixes, I'd be happy to learn.

**Pull request merged**